pkg/upgrade/upgrade.go

```diff
@@ -0,0 +1,375 @@
+
+	original, err := json.MarshalIndent(clusterConfig, "", "  ")
+	if err != nil {
+		return fmt.Errorf("marshal existing clusterconfig: %w", err)
+	}
+
+	modified, err := json.MarshalIndent(desired, "", "  ")
+	if err != nil {
+		return fmt.Errorf("marshal desired clusterconfig: %w", err)
+	}
+
+	patchData, err := jsonpatch.CreateMergePatch(original, modified)
+	if err != nil {
+		return fmt.Errorf("create json merge patch: %w", err)
+	}
+
+	if string(patchData) == "{}" {
+		log.Info("K0s cluster config already patched")
+		return nil
+	}
+
+	log.V(2).Info("Patching K0s cluster config with merge patch", "patch", string(patchData))
+
+	patch := client.RawPatch(types.MergePatchType, patchData)
+	err = cli.Patch(ctx, clusterConfig, patch)
+	if err != nil {
+		return fmt.Errorf("patch clusterconfig: %w", err)
+	}
+
+	log.Info("K0s cluster config patched")
+
+	return nil
+}
+
+func setClusterConfigOperatorChart(clusterConfig *k0sv1beta1.ClusterConfig, operatorChart k0sv1beta1.Chart) *k0sv1beta1.ClusterConfig {
+	desired := clusterConfig.DeepCopy()
+	if desired.Spec == nil {
+		desired.Spec = &k0sv1beta1.ClusterSpec{}
+	}
+	if desired.Spec.Extensions == nil {
+		desired.Spec.Extensions = &k0sv1beta1.ClusterExtensions{}
+	}
+	if desired.Spec.Extensions.Helm == nil {
+		desired.Spec.Extensions.Helm = &k0sv1beta1.HelmExtensions{}
+	}
+	for i, chart := range desired.Spec.Extensions.Helm.Charts {
+		if chart.Name == operatorChartName {
+			desired.Spec.Extensions.Helm.Charts[i] = operatorChart
+			return desired
+		}
+	}
+	desired.Spec.Extensions.Helm.Charts = append(desired.Spec.Extensions.Helm.Charts, operatorChart)
+	return desired
+}
+
+func getExistingClusterConfig(ctx context.Context, cli client.Client) (*k0sv1beta1.ClusterConfig, error) {
+	clusterConfig := &k0sv1beta1.ClusterConfig{}
+	err := cli.Get(ctx, client.ObjectKey{Name: clusterConfigName, Namespace: clusterConfigNamespace}, clusterConfig)
+	if err != nil {
+		return nil, fmt.Errorf("get chart: %w", err)
+	}
+	return clusterConfig, nil
+}
+
+func getOperatorChartFromMetadata(metadata *ectypes.ReleaseMetadata) (k0sv1beta1.Chart, error) {
+	for _, chart := range metadata.Configs.Charts {
+		if chart.Name == operatorChartName {
+			return chart, nil
+		}
+	}
+	return k0sv1beta1.Chart{}, fmt.Errorf("chart not found")
+}
+
+const (
+	installationNameAnnotation = "embedded-cluster.replicated.com/installation-name"
+)
+
+func airgapDistributeArtifacts(ctx context.Context, cli client.Client, in *clusterv1beta1.Installation, localArtifactMirrorImage string) error {
+	// in airgap installations let's make sure all assets have been copied to nodes.
+	// this may take some time so we only move forward when 'ready'.
+	err := ensureAirgapArtifactsOnNodes(ctx, cli, in, localArtifactMirrorImage)
+	if err != nil {
+		return fmt.Errorf("ensure airgap artifacts: %w", err)
+	}
+
+	// once all assets are in place we can create the autopilot plan to push the images to the
+	// local registry.
```

Comment:
nit: this k0s airgap plan actually loads images into containerd, not the local registry.

Reply:
updated